### PR TITLE
reports: include "other info" per alert instance in reports

### PR DESCRIPTION
**Pull request by aryangupta701**
Fixes zaproxy/zaproxy#7260

Earlier the Tradition JSON Plus report was identifying the otherinfo of every alert as same. However it is not always the case.
The Traditional JSON Plus report now identifies the "otherinfo" of each alert as different.

### CODE REVIEW

File: addOns/reports/CHANGELOG.md
Comment (kingthorin) [inline addOns/reports/CHANGELOG.md:9]: > - The Traditional JSON Plus report now treats "otherinfo" as different per alert (Issue 7260).

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/main/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

**Comment by kingthorin** [comment]
Please update the PR comment to close the issue once merged. https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests

**Comment by kingthorin** [comment]
Looks like the code needs to have our standard formatting applied.

`./gradlew :aO:reports:spotlessApply`

Add, commit (amending), push (force).

**Comment by aryangupta701** [comment]
Okk

**Comment by kingthorin** [comment]
Why?

**Comment by aryangupta701** [comment]
sorry, I was having some confusion using github, now its fine, I got it.

**Comment by kingthorin** [comment]
In the future it'll make your life easier if you create specific branches for PRs/features/issues instead of working on `main`.

**Comment by aryangupta701** [comment]
Yes, in future I will work on different branch. Thankyou

File: addOns/reports/CHANGELOG.md
Comment (thc202) [reply addOns/reports/CHANGELOG.md]: The name of the report is `Traditional JSON Report with requests and responses` though.

Comment (aryangupta701) [reply addOns/reports/CHANGELOG.md]: Yes, but that is only in the GUI, rest everywhere the name is `Traditional JSON Plus Report`. Should I change it ?

Comment (thc202) [reply addOns/reports/CHANGELOG.md]: Do you mean `traditional-json-plus`? That's the ID of the report not its name.

Comment (aryangupta701) [reply addOns/reports/CHANGELOG.md]: Oh okay. Should I make any changes?

Comment (thc202) [reply addOns/reports/CHANGELOG.md]: It would be better to use the correct name.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: This should not duplicate everything, extract a common method and have only the specific `otherinfo` checked in each method.

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: Still I think we should have consistent handling of the other info in all reports (I didn't check all but the other JSON is not correct).

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: yes, correct. I have only changed it for traditional-json-plus. I confirmed about that from Simon.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: ok I will make a common method for that

Comment (kingthorin) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: Simon did suggest starting with this one :wink:

@thc202 wouldn’t this be a breaking change for the actual traditional reports? (Not that it shouldn’t be done to the other non-trads, just want to land in the right place).

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: Is the confirmation the comment in the issue or was that a direct message?

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: yes in the issue.

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: Only if we break it, we can keep adding the otherinfo field as we do now.

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: https://github.com/zaproxy/zaproxy/issues/7260#issuecomment-1406385828 is this one? That seems to be a suggestion to get to know the code. Anyway, we should not fix the issue if we are going to merge just these changes IMO.

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: fwiw this is a breaking change for this report.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: I think you are correct. Please tell me what should I do now ? Should I change all the reports ?

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: @psiinon can you please finalize what should be done ?

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: Up to you to decide if you want to change them all (though first we all should agree that's what should be done).

Re the report being changed now, we should not break it, the other info in the general details should be brought back. The tests should also have at least two alert instances to verify that the other info is being added correctly.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: I checked all the reports and found that only these reports are using "otherinfo" and only these will be needed to be updated: 

1. Traditional JSON report 
2. Traditional JSON report with requests and responses
3. Traditional XML report 
4. Traditional JSON report with requests and responses

Tell me if any other report also need to be changed.

Tell me when you all agree to make changes in these reports, then I will start the work 😄.

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: That's bad, all reports should include the other info. That list is not complete though `SARIF` and `Risk and Confidence HTML` include other info as well.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: I cannot find where `SARIF` and `Risk and Confidence HTML` are using other info. Can you help me find it please ? Here I have included the snapshots of both the reports : 

Screenshot from 2023-01-29 15-45-24
Screenshot from 2023-01-29 15-46-38

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: @thc202 there are no tests written for xml-plus report should I write that?

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: Ideally we should have tests for everything. Again, it's up to you to decide that (I don't want to push work).

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: okk I will write that.

File: addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-traditional-json.html
Comment (kingthorin) [inline addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-traditional-json.html:53]: This should be restored. Although we know it isn't technically correct, removing it may break handling/processing that existing users have in place.

Comment (aryangupta701) [reply addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-traditional-json.html:53]: okay I will correct that eveywhere.

Comment (kingthorin) [reply addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-traditional-json.html:53]: @thc202 that is what you meant right?

Comment (thc202) [reply addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-traditional-json.html:53]: Yes.

File: addOns/reports/CHANGELOG.md
Comment (thc202) [inline addOns/reports/CHANGELOG.md:8]: There's already a `Changed` section although I would mention this under `Added`, since it's a new "field".
Also good to mention under `Deprecated` section that the "old" other info entry is to be phased out.

File: addOns/reports/src/test/resources/org/zaproxy/addon/reports/resources/basic-traditional-xml-plus.xml
Comment (thc202) [inline addOns/reports/src/test/resources/org/zaproxy/addon/reports/resources/basic-traditional-xml-plus.xml:34]: This should be reverted.

File: addOns/reports/CHANGELOG.md
Comment (aryangupta701) [reply addOns/reports/CHANGELOG.md:8]: okk, please tell me how to write tell in more details.

File: addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-traditional-json.html
Comment (kingthorin) [reply addOns/reports/src/main/javahelp/org/zaproxy/addon/reports/resources/help/contents/report-traditional-json.html:53]: This still needs to be addressed

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (kingthorin) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:721]: ?

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (kingthorin) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:868]: checkXmlPlus

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (kingthorin) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:1059]: (!isXmlPlus)

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:868]: the Boolean variable name or the function name ? Actually this functions checks for both Xml and Xml-Plus. If isXMLPlus is true then it check for xml-plus else for xml.

Comment (kingthorin) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:868]: Both should follow camelCaps

**Comment by aryangupta701** [comment]
I messed up something. fixing..

**Comment by aryangupta701** [comment]
Screenshot from 2023-01-31 18-56-20
@kingthorin I mistakenly signoff the commit made by you. I am not able to correct it. Please help me.

**Comment by kingthorin** [comment]
I can put together some instructions but won't get to it till this evening. In the mean time you could look into 'interactive rebase', you'd `skip` my commit and `pick` yours (or even set yours are `fixup`).

**Comment by aryangupta701** [comment]
okk

**Comment by kingthorin** [comment]
I had put this together quite a while ago. You're probably not using Eclipse but it might provide some hints that would help: https://www.instructables.com/How-to-Rebase-Interactive-With-Eclipse-eGit/

**Comment by aryangupta701** [comment]
Thankyou

**Comment by aryangupta701** [comment]
Thankyou so much @kingthorin

File: addOns/reports/CHANGELOG.md
Comment (kingthorin) [inline addOns/reports/CHANGELOG.md:16]:
```suggestion

### Deprecated
- The old "otherinfo" entry to be phased out.

### Added
- A description of riskdesc fields in the relevant report templates' help (Issue 7445).
- The Traditional JSON report, Traditional JSON Report with requests and responses, Traditional XML Report and Traditional XML Report with requests and responses now has "otherinfo" field per alert (Issue 7260).

```

Comment (kingthorin) [reply addOns/reports/CHANGELOG.md:16]: May as well combine the two additions under a single heading :wink:

**Comment by kingthorin** [comment]
No problem, happy to help.

File: addOns/reports/CHANGELOG.md
Comment (kingthorin) [inline addOns/reports/CHANGELOG.md:10]:
```suggestion
### Added 
- The Traditional JSON report, Traditional JSON Report with requests and responses, Traditional XML Report and Traditional XML Report with requests and responses now has "otherinfo" field per alert (Issue 7260).
```

Comment (kingthorin) [inline addOns/reports/CHANGELOG.md:10]: This blank can go, so that the two bullets are after one another.

File: addOns/reports/CHANGELOG.md
Comment (kingthorin) [inline addOns/reports/CHANGELOG.md:17]: The heading and one blank line should be removed.

Comment (aryangupta701) [reply addOns/reports/CHANGELOG.md:17]: yes sorry

**Comment by aryangupta701** [comment]
@kingthorin Is there any need for some more changes? Please verify.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (kingthorin) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:747]: JSONObject is outside our control but for our local variables/methods we should use proper camelCaps. So `isJsonPlus` here and at 793, etc.

**Review by kingthorin — COMMENTED** [review]
I only see one minor thing.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:747]: oh okay. I will change that.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java]: done

**Review by kingthorin — APPROVED** [review]
Thanks LGTM

**Comment by aryangupta701** [comment]
@thc202 Is there any need for some more changes? Please review.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java]: This was not yet addressed: https://github.com/zaproxy/zap-extensions/pull/4381#discussion_r1089915714

File: addOns/reports/CHANGELOG.md
Comment (thc202) [inline addOns/reports/CHANGELOG.md:9]: `per alert instance`

File: addOns/reports/CHANGELOG.md
Comment (thc202) [inline addOns/reports/CHANGELOG.md:13]: This should be more clear, e.g. `The "otherinfo" under the alert will be removed, superseded by "otherinfo" under the alert instances.`

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:866]: - `private static`
- `boolean`

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:747]: - `private static`
- `boolean`

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:793]: Add braces.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:1057]: Add braces.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java]: you mean the two alert instances part ?

Comment (kingthorin) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java]: Thank you both, sorry I missed it.

**Comment by aryangupta701** [comment]
okk I will correct these all

**Comment by aryangupta701** [comment]
why there is a branch conflict here ?

**Comment by aryangupta701** [comment]
oh I got it. I merged a new version which had some unreleased changes in the released part.

**Comment by kingthorin** [comment]
Because reports was released and the CHANGELOG updated in the meantime.

(For future reference If you rebase instead of merging, then force push your PR branch will be cleaner.)

To address the conflict you just need to ensure your CHANGELOG entry is actually ordered properly and includes the new content. Ex: Everything in this screenshot is actually fine (just check the content & ordering, add it, and commit it):
image

**Comment by aryangupta701** [comment]
Thankyou

**Comment by aryangupta701** [comment]
@thc202 Could you please review this Pull Request and let me know if there are any necessary changes to be made?

File: addOns/reports/CHANGELOG.md
Comment (thc202) [inline addOns/reports/CHANGELOG.md:10]: This is not relevant for the user.

Comment (kingthorin) [reply addOns/reports/CHANGELOG.md:10]: The second bullet could just be dropped. Agreed this was the unit test change that you implemented but unit test stuff isn't relevant to end users.

Comment (aryangupta701) [reply addOns/reports/CHANGELOG.md:10]: okk sure

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:794]: Split the method (e.g. `checkJsonAlertInstance(…)` and `checkJsonAlertInstanceAndMessages(…)`) and call each as required (boolean should no longer be needed here).

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:669]: Use a different alert with different other info.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:1018]: Same here to split.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:1104]: Should be moved to `// Then`, same in following test.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:1104]: okk working on all

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:669]: do you mean there should be two alerts with one instance each ?

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:669]: An alert is an instance, yes.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:669]: I am still not clear. 
1. 
Screenshot from 2023-02-22 10-28-35


2. 
Screenshot from 2023-02-22 10-27-31

Please tell out of these which is correct.

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:669]: The second example.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:1018]: For Json and JsonPlus I have splitted. But it will be very messy if I do the same with xml and xmlPlus because `isXmlPlus` is  being used at many random places.

**Comment by aryangupta701** [comment]
@thc202 Please let me know if there is anything else which needs to be changed

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:1112]: This should be kept `When`, it's the other statements that need to be moved to under `Then`.

File: addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java
Comment (thc202) [inline addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:672]: This does not look right.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:672]: The method or the way of writing ?

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:672]: Calling it twice, the second call overrides the first.

Comment (aryangupta701) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:672]: I think only one of these statements should be written

Comment (thc202) [reply addOns/reports/src/test/java/org/zaproxy/addon/reports/ExtensionReportsUnitTest.java:672]: Right.

**Comment by aryangupta701** [comment]
@thc202 please let me know if anything needs to be changed.

**Comment by thc202** [comment]
Just fixing up the commits, I think.

**Comment by kingthorin** [comment]
You still have an unnecessary commit.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Comment by aryangupta701** [comment]
Thank you for your support @kingthorin @thc202 :)

**Comment by kingthorin** [comment]
No problem, thanks for sticking with it!

**Pull request merged**
